src/win.c

```diff
@@ -2162,6 +2153,58 @@ void map_win_start(session_t *ps, struct managed_win *w) {
 	}
 }
 
+/**
+ * Update target window opacity depending on the current state.
+ */
+void win_update_opacity_target(session_t *ps, struct managed_win *w) {
+	auto opacity_target_old = w->opacity_target;
+	w->opacity_target = win_calc_opacity_target(ps, w, false);
+
+	if (opacity_target_old != w->opacity_target) {
+		if (w->state == WSTATE_MAPPED) {
+			// Opacity target changed while MAPPED. Transition to FADING.
+			assert(w->opacity == opacity_target_old);
+			w->opacity_target_old = opacity_target_old;
+			w->state = WSTATE_FADING;
+			log_debug("Window %#010x (%s) opacity %f, opacity target %f, set "
+			          "old target %f",
+			          w->base.id, w->name, w->opacity, w->opacity_target,
+			          w->opacity_target_old);
+			if (!ps->redirected) {
+				CHECK(!win_skip_fading(ps, w));
+			}
+		} else if (w->state == WSTATE_MAPPING) {
+			// Opacity target changed while fading in.
+			if (w->opacity >= w->opacity_target) {
+				// Already reached new target opacity. Transition to
+				// FADING.
+				map_win_finish(w);
+				w->opacity_target_old = fmax(opacity_target_old, w->opacity);
+				w->state = WSTATE_FADING;
+				log_debug("Window %#010x (%s) opacity %f already reached "
+				          "new opacity target %f while mapping, set old "
+				          "target %f",
+				          w->base.id, w->name, w->opacity,
+				          w->opacity_target, w->opacity_target_old);
+				if (!ps->redirected) {
+					CHECK(!win_skip_fading(ps, w));
+				}
+			}
+		} else if (w->state == WSTATE_FADING) {
+			// Opacity target changed while FADING.
+			if (w->opacity < opacity_target_old && w->opacity > w->opacity_target) {
+				// Changed while fading in and will fade out.
+				w->opacity_target_old = fmax(opacity_target_old, w->opacity);
```

Comment:
We get here if `w->opacity < opacity_target_old`, so this `fmax` will always return `opacity_target_old`?

Reply:
That's true. In fact, `w->opacity_target_old` should be updated in the opposite case (fade-out to fade-in) as well to enable smooth fading in this case as well. The latter does however not affect the issue of exploding `blur_opacity`.